### ISSUE TITLE
Corresponding DropdownButtonFormField throws exception that DropDownButton does not experience

### ISSUE THREAD

**Issue by davidferraro** [flutter/flutter#64820]
When creating corresponding dropdowns (the items in the second dropdown change when the first dropdown is changed), I am experiencing a crash when using the DropdownButtonFormField widget. The same code attached changed to the DropDownButton widget works as expected.

## Steps to Reproduce

1. Run `flutter create app`.
2. Update the main.dart file to:
```
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class City {
  int cityId;
  String cityName;
  City({this.cityId, this.cityName});
}

class CountryState {
  int stateId;
  String stateName;
  List<City> cities;
  CountryState({this.stateId, this.stateName, this.cities});
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Bug Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<CountryState> states = [
    CountryState(stateId: 1, stateName: "New York", cities: [
      City(cityId: 1, cityName: "New York City"),
      City(cityId: 2, cityName: "Buffalo")
    ]),
    CountryState(stateId: 2, stateName: "California", cities: [
      City(cityId: 3, cityName: "Los Angeles"),
      City(cityId: 4, cityName: "San Diego")
    ])
  ];

  int selectedStateId;
  int selectedCityId;

  @override
  void initState() {
    selectedStateId = states[0].stateId;
    selectedCityId = states[0].cities[0].cityId;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            DropdownButtonFormField<int>(
              decoration: InputDecoration(labelText: 'State'),
              value: selectedStateId,
              items: states.map((CountryState state) {
                return DropdownMenuItem<int>(
                  value: state.stateId,
                  child: Text(states
                      .singleWhere((x) => x.stateId == state.stateId)
                      .stateName),
                );
              }).toList(),
              onChanged: (int newStateId) {
                setState(() {
                  this.selectedCityId = states
                      .singleWhere((x) => x.stateId == newStateId)
                      .cities[0]
                      .cityId;
                  this.selectedStateId = newStateId;
                });
              },
            ),
            DropdownButtonFormField<int>(
              decoration: InputDecoration(labelText: 'City'),
              value: selectedCityId,
              items: states
                  .singleWhere((x) => x.stateId == selectedStateId)
                  .cities
                  .map((City city) {
                return DropdownMenuItem<int>(
                  value: city.cityId,
                  child: Text(states
                      .singleWhere((x) => x.stateId == selectedStateId)
                      .cities
                      .singleWhere((x) => x.cityId == city.cityId)
                      .cityName),
                );
              }).toList(),
              onChanged: (int newCityId) {
                setState(() {
                  this.selectedCityId = newCityId;
                });
              },
            )
          ],
        ),
      ),
    );
  }
}

```
3. Change the value of the first dropdown.

**Expected results:** The items in the city dropdown should update to those for that state.

**Actual results:** The following exception is thrown:
```There should be exactly one item with [DropdownButton]'s value: 2. 
Either zero or 2 or more [DropdownMenuItem]s were detected with the same value
'package:flutter/src/material/dropdown.dart':
Failed assertion: line 827 pos 15: 'items == null || items.isEmpty || value == null ||
              items.where((DropdownMenuItem<T> item) {
                return item.value == value;
              }).length == 1'
```

<details>
  <summary>Logs</summary>

```
[+23709 ms] I/flutter (25451): ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
[   +8 ms] I/flutter (25451): The following assertion was thrown building DropdownButtonFormField<int>(dirty, dependencies:
[  +63 ms] I/flutter (25451): [_LocalizationsScope-[GlobalKey#bddc5], _InheritedTheme], state:
[   +3 ms] I/flutter (25451): _DropdownButtonFormFieldState<int>#6ed50):
[   +4 ms] I/flutter (25451): There should be exactly one item with [DropdownButton]'s value: 1.
[   +3 ms] I/flutter (25451): Either zero or 2 or more [DropdownMenuItem]s were detected with the same value
[  +74 ms] I/flutter (25451): 'package:flutter/src/material/dropdown.dart':
[   +3 ms] I/flutter (25451): Failed assertion: line 827 pos 15: 'items == null || items.isEmpty || value == null ||
[   +3 ms] I/flutter (25451):               items.where((DropdownMenuItem<T> item) {
[  +50 ms] I/flutter (25451):                 return item.value == value;
[   +3 ms] I/flutter (25451):               }).length == 1'
[   +4 ms] I/flutter (25451): 
[   +5 ms] I/flutter (25451): Either the assertion indicates an error in the framework itself, or we should provide substantially
[  +50 ms] I/flutter (25451): more information in this error message to help you determine and fix the underlying cause.
[   +3 ms] I/flutter (25451): In either case, please report this assertion by filing a bug on GitHub:
[   +5 ms] I/flutter (25451):   https://github.com/flutter/flutter/issues/new?template=BUG.md
[   +3 ms] I/flutter (25451): 
[  +72 ms] I/flutter (25451): The relevant error-causing widget was:
[   +4 ms] I/flutter (25451):   DropdownButtonFormField<int>
[   +3 ms] I/flutter (25451):   file:///C:/Projects/bugs/dropdown/dropdownformfield_bug/lib/main.dart:97:13
[   +3 ms] I/flutter (25451): 
[ +102 ms] I/flutter (25451): When the exception was thrown, this was the stack:
[  +53 ms] I/flutter (25451): #2      new DropdownButton (package:flutter/src/material/dropdown.dart:827:15)
[   +2 ms] I/flutter (25451): #3      new DropdownButtonFormField.<anonymous closure> (package:flutter/src/material/dropdown.dart:1493:23)
[   +1 ms] I/flutter (25451): #4      FormFieldState.build (package:flutter/src/widgets/form.dart:437:26)
[   +2 ms] I/flutter (25451): #5      StatefulElement.build (package:flutter/src/widgets/framework.dart:4619:28)
[   +1 ms] I/flutter (25451): #6      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4502:15)
[   +3 ms] I/flutter (25451): #7      StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4675:11)
[  +12 ms] I/flutter (25451): #8      Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[   +1 ms] I/flutter (25451): #9      StatefulElement.update (package:flutter/src/widgets/framework.dart:4707:5)
[   +1 ms] I/flutter (25451): #10     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +1 ms] I/flutter (25451): #11     RenderObjectElement.updateChildren (package:flutter/src/widgets/framework.dart:5522:32)
[   +1 ms] I/flutter (25451): #12     MultiChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5957:17)
[  +16 ms] I/flutter (25451): #13     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +1 ms] I/flutter (25451): #14     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5837:14)
[   +1 ms] I/flutter (25451): #15     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +1 ms] I/flutter (25451): #16     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[  +16 ms] I/flutter (25451): #17     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[   +1 ms] I/flutter (25451): #18     StatelessElement.update (package:flutter/src/widgets/framework.dart:4583:5)
[   +1 ms] I/flutter (25451): #19     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +1 ms] I/flutter (25451): #20     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[  +15 ms] I/flutter (25451): #21     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[   +1 ms] I/flutter (25451): #22     ProxyElement.update (package:flutter/src/widgets/framework.dart:4862:5)
[   +1 ms] I/flutter (25451): #23     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +1 ms] I/flutter (25451): #24     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[  +15 ms] I/flutter (25451): #25     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[   +1 ms] I/flutter (25451): #26     ProxyElement.update (package:flutter/src/widgets/framework.dart:4862:5)
[   +1 ms] I/flutter (25451): #27     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +2 ms] I/flutter (25451): #28     RenderObjectElement.updateChildren (package:flutter/src/widgets/framework.dart:5522:32)
[  +16 ms] I/flutter (25451): #29     MultiChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5957:17)
[   +2 ms] I/flutter (25451): #30     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[  +38 ms] I/flutter (25451): #31     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[  +59 ms] I/flutter (25451): #32     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4675:11)
[   +3 ms] I/flutter (25451): #33     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[  +37 ms] I/flutter (25451): #34     StatefulElement.update (package:flutter/src/widgets/framework.dart:4707:5)
[   +6 ms] I/flutter (25451): #35     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[  +38 ms] I/flutter (25451): #36     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[   +3 ms] I/flutter (25451): #37     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[  +37 ms] I/flutter (25451): #38     ProxyElement.update (package:flutter/src/widgets/framework.dart:4862:5)
[   +3 ms] I/flutter (25451): #39     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[  +37 ms] I/flutter (25451): #40     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[   +3 ms] I/flutter (25451): #41     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4675:11)
[  +39 ms] I/flutter (25451): #42     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[   +3 ms] I/flutter (25451): #43     StatefulElement.update (package:flutter/src/widgets/framework.dart:4707:5)
[  +36 ms] I/flutter (25451): #44     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +3 ms] I/flutter (25451): #45     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5837:14)
[  +59 ms] I/flutter (25451): #46     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[  +60 ms] I/flutter (25451): #47     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[   +3 ms] I/flutter (25451): #48     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[  +44 ms] I/flutter (25451): #49     StatelessElement.update (package:flutter/src/widgets/framework.dart:4583:5)
[  +37 ms] I/flutter (25451): #50     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +3 ms] I/flutter (25451): #51     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5837:14)
[  +33 ms] I/flutter (25451): #52     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +2 ms] I/flutter (25451): #53     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[   +2 ms] I/flutter (25451): #54     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4675:11)
[   +2 ms] I/flutter (25451): #55     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[  +18 ms] I/flutter (25451): #56     StatefulElement.update (package:flutter/src/widgets/framework.dart:4707:5)
[   +2 ms] I/flutter (25451): #57     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[  +18 ms] I/flutter (25451): #58     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[   +2 ms] I/flutter (25451): #59     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4675:11)
[   +1 ms] I/flutter (25451): #60     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[  +19 ms] I/flutter (25451): #61     StatefulElement.update (package:flutter/src/widgets/framework.dart:4707:5)
[  +54 ms] I/flutter (25451): #62     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +2 ms] I/flutter (25451): #63     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[   +2 ms] I/flutter (25451): #64     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[  +30 ms] I/flutter (25451): #65     ProxyElement.update (package:flutter/src/widgets/framework.dart:4862:5)
[   +4 ms] I/flutter (25451): #66     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +2 ms] I/flutter (25451): #67     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[   +3 ms] I/flutter (25451): #68     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[   +3 ms] I/flutter (25451): #69     ProxyElement.update (package:flutter/src/widgets/framework.dart:4862:5)
[  +34 ms] I/flutter (25451): #70     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +2 ms] I/flutter (25451): #71     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[   +2 ms] I/flutter (25451): #72     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4675:11)
[  +27 ms] I/flutter (25451): #73     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[   +5 ms] I/flutter (25451): #74     StatefulElement.update (package:flutter/src/widgets/framework.dart:4707:5)
[   +2 ms] I/flutter (25451): #75     Element.updateChild (package:flutter/src/widgets/framework.dart:3201:15)
[   +3 ms] I/flutter (25451): #76     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4527:16)
[   +2 ms] I/flutter (25451): #77     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4675:11)
[   +2 ms] I/flutter (25451): #78     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
[  +29 ms] I/flutter (25451): #79     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2627:33)
[   +3 ms] I/flutter (25451): #80     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:883:20)
[   +3 ms] I/flutter (25451): #81     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:284:5)
[   +3 ms] I/flutter (25451): #82     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1113:15)
[  +37 ms] I/flutter (25451): #83     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1052:9)
[   +3 ms] I/flutter (25451): #84     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:968:5)
[   +3 ms] I/flutter (25451): #88     _invoke (dart:ui/hooks.dart:261:10)
[  +35 ms] I/flutter (25451): #89     _drawFrame (dart:ui/hooks.dart:219:3)
[   +2 ms] I/flutter (25451): (elided 5 frames from class _AssertionError and dart:async)
[   +3 ms] I/flutter (25451): 
[  +34 ms] I/flutter (25451): ════════════════════════════════════════════════════════════════════════════════════════════════════
[   +4 ms] I/flutter (25451): Another exception was thrown: A RenderFlex overflowed by 99335 pixels on the bottom.
```

```
[√] Flutter (Channel stable, v1.17.1, on Microsoft Windows [Version 10.0.18363.1016], locale en-US)
    • Flutter version 1.17.1 at C:\src\flutter
    • Framework revision f7a6a7906b (4 months ago), 2020-05-12 18:39:00 -0700
    • Engine revision 6bc433c6b6
    • Dart version 2.8.2

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at C:\Users\dferraro\AppData\Local\Android\sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Android Studio (version 4.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code (version 1.48.1)
    • VS Code at C:\Users\dferraro\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.12.2

[√] Connected device (1 available)
    • Pixel 4a • 08041JEC203120 • android-arm64 • Android 10 (API 29)

• No issues found!
```

</details>

**Comment by darshankawar** [comment]
Hi @davidferraro,
I tried your code on latest Stable (1.20.2) and didn't see the exception.

[64820.mp4.zip](https://github.com/flutter/flutter/files/5148603/64820.mp4.zip)

Also, the exception you were getting might have arised due to the duplicated value you may be passing. 

Please see https://flutter.dev/community for resources and asking questions like this.
you may also get some help if you post it on Stack Overflow and if you need help with your code, please see https://www.reddit.com/r/flutterhelp/
Closing, as this isn't an issue with Flutter itself. If you disagree, please write in the comments and I will reopen it.
Thank you

**Comment by davidferraro** [comment]
Hi @darshankawar ,

I updated to 1.20.2 and my sample code now no longer displays this behavior. I believe whatever issue with the DropdownButtonFormField I was experiencing must have been fixed since 1.17.1 - cheers!
